CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/sna/centrality/PagerankCentralityPlugin.java

```diff
@@ -95,118 +92,181 @@ public PluginParameters createParameters() {
-                                    if ((treatUndirectedBidirectional && transactionDirection == GraphConstants.FLAT)
-                                            || (neighbourId < otherId && transactionDirection == GraphConstants.UPHILL)
-                                            || (neighbourId > otherId && transactionDirection == GraphConstants.DOWNHILL)) {
-                                        neighbourOutCount++;
-                                    }
-                                }
-                            }
+        }
+
+        // update the graph with pagerank values
+        PagerankVertex.commitPageranksToGraph();
+    }
+    
+    private static class PagerankVertex {
+        private static GraphWriteMethods graph;
+        private static final Set<PagerankVertex> sinks = new HashSet<>();
+        private static boolean treatUndirectedBidirectional;
+        private static boolean normaliseByAvailable;
+        private static double dampingFactor = 0;
+        private static int vertexCount = 0;
+        private static PagerankVertex[] pagerankVertices;
+        private static double sinkPagerankContribution = 0;
+        private static double delta = 0;
+        private static int pagerankAttribute;
+        private static double baseContribution;
+        private final int vertexId;
+        private final Set<PagerankVertex> neighbours = new HashSet<>();
+        private int outCount = 0;
+        private double pagerank;
+        private double pagerankContribution;
+        private double stagedPagerank;
+        private boolean isSink = false;
 
-                            vertexInLinks.get(vertexId).add(Tuple.create(neighbourId, neighbourOutCount));
+        private PagerankVertex(final int vertexPosition) {
+            this.vertexId = graph.getVertex(vertexPosition);
+            this.pagerank = 1.0/vertexCount;
+            this.pagerankContribution = 0;
+        }
+        
+        public static void initialiseAllPagerankVertices(final GraphWriteMethods graph, final boolean treatUndirectedBidirectional, final double dampingFactor, final boolean normaliseByAvailable) {
+            PagerankVertex.graph = graph;
+            PagerankVertex.vertexCount = graph.getVertexCount();
+            PagerankVertex.treatUndirectedBidirectional = treatUndirectedBidirectional;
+            PagerankVertex.normaliseByAvailable = normaliseByAvailable;
+            PagerankVertex.pagerankAttribute = PAGERANK_ATTRIBUTE.ensure(graph);
+            PagerankVertex.pagerankVertices =  new PagerankVertex[vertexCount];
+            PagerankVertex.dampingFactor = dampingFactor;
+            PagerankVertex.baseContribution = (1 - dampingFactor) / vertexCount;
+            
+            for (int vertexPosition = 0; vertexPosition < vertexCount; vertexPosition++) {
+                // For every vertex
+                PagerankVertex prVertex = new PagerankVertex(vertexPosition);
```

Comment:
Is there a reason for instantiating PagerankVertex with a vertex position over a vertex id? I can't see this causing any issues in this instance, but it's generally not recommended practice due to the volatility of vertex position.

Reply:
Hmm, I think the reason is a combination of lines 161 and 192. Im using the position just so that I can add it to the pageRankVerticies array in that position and then refer back to it.

If I instead stores them in a Set and used the id as the key then I could change line 192 to retrieve them by ID instead and not use the position at all. Do you think that would be better?